wandb/sdk/launch/runner/kubernetes_runner.py

```diff
@@ -64,11 +66,166 @@
+            and container_status.state.waiting.reason == "ContainerCreating"
+        ):
+            return True
+    return False
+
+
+class KubernetesRunMonitor:
+    def __init__(
+        self,
+        job_field_selector: str,
+        pod_label_selector: str,
+        namespace: str,
+        batch_api: "BatchV1Api",
+        core_api: "CoreV1Api",
+    ) -> None:
+        """Initial KubernetesRunMonitor.
+
+        Arguments:
+            jobname: Name of the job.
+
+        Returns:
+            None.
+        """
+        self.pod_label_selector = pod_label_selector
+        self.job_field_selector = job_field_selector
+        self.namespace = namespace
+        self.batch_api = batch_api
+        self.core_api = core_api
+
+        self._status_lock = Lock()
+        self._status = Status("starting")
+
+        self._watch_job_thread = Thread(target=self._watch_job, daemon=True)
+        self._watch_pods_thread = Thread(target=self._watch_pods, daemon=True)
+
+        self._job_watcher = watch.Watch()
+        self._pod_watcher = watch.Watch()
+
+    def start(self) -> None:
+        """Start the run monitor."""
+        if self._watch_job_thread.is_alive() or self._watch_pods_thread.is_alive():
+            raise LaunchError(
+                "Attempted to start monitor that has already started"
+            )  # TODO: what should I do here?
+        self._watch_job_thread.start()
+        self._watch_pods_thread.start()
+
+    def stop(self) -> None:
+        """Stop the run monitor."""
+        self._job_watcher.stop()
+        self._pod_watcher.stop()
+
+    def _set_status(self, status: Status) -> None:
+        """Set the run status."""
+        with self._status_lock:
+            self._status = status
+
+    def get_status(self) -> Status:
+        """Get the run status."""
+        with self._status_lock:
+            return self._status
+
+    def _watch_pods(self) -> None:
+        """Watch for pods created matching the jobname."""
+        try:
+            # Stream with no timeout polling for pod status updates
+            for event in self._pod_watcher.stream(
+                self.core_api.list_namespaced_pod,
+                namespace=self.namespace,
+                label_selector=self.pod_label_selector,
+            ):
+                type = event.get("type")
+                object = event.get("object")
+
+                if type == "MODIFIED":
+                    if object.status.phase == "Running":
+                        self._set_status(Status("running"))
+                if _is_preempted(object.status):
+                    self._set_status(Status("preempted"))
+                    self.stop()
+                    break
+                if _is_container_creating(object.status):
+                    self._set_status(Status("starting"))
+
+        # This can happen if the connection is lost to the Kubernetes API server
+        # and cannot be re-established.
+        except ApiException as e:
```

Comment:
Both exception types have the same comment, is that correct?

Also, should we really raise here? Is there no case where the watch api recovers from these exceptions?

Reply:
These only happen if the kubernetes sdk fails it's retry loop, which should really mean that the stream is dead because the object is gone or something else is wrong with the cluster. If we can't see the job anymore I think we should error and mark it failed righT?